hibernate-core/src/main/java/org/hibernate/type/TypeHelper.java

```diff
@@ -192,7 +192,10 @@ public static Object[] replaceAssociations(
 			}
 			else {
 				final Type type = types[i];
-				if ( type.isComponentType() ) {
+				if (!type.isMutable()) {
+					copied[i] = types[i].replace( currentOriginal, target[i], session, owner, copyCache, foreignKeyDirection );
+				}
```

Comment:
This change doesn't look quite right to me. It means that every non-mutable type gets `replace()`d, but I believe this method is only supposed to `replace()` associations.

That change isn't likely to cause any test failures, but it results in unnecessary work.

Reply:
Then I will change that first if to:

```java
				if ( type.isComponentType() && !type.isMutable() ) {
					copied[i] = target[i];
				}
```

Reply:
Actually, if I'm not mistaken, what you really need is to do the same thing for record types that we currently do for `CompositeUserType`s.

But before anything else, _pull_, I just merged something that should make this much easier to solve.

Reply:
Unfortunately, I was too fast, messef something. And when tried to fix things, I've got message saying "Pull request successfully merged and closed".Closed yes, merged no. Anyway, it seems that this MR is now dead.

Reply:
> Actually, if I'm not mistaken, what you really need is to do the same thing for record types that we currently do for `CompositeUserType`s.

There is no special record type, it is hus ComponentType